[JSC] Fix DFG::AbstractHeap initialization from pointer on armv7

eugeneia opened the pull request:
#### 69218f370db8606813041a0efeddfcab5643cc1c
<pre>
[JSC] Fix DFG::AbstractHeap initialization from pointer on armv7
<a href="https://bugs.webkit.org/show_bug.cgi?id=281378">https://bugs.webkit.org/show_bug.cgi?id=281378</a>

Reviewed by Justin Michaud.

The previous incantation would cause 32 bit pointers to be
sign extended into a int64_t slot. We want to widen then
value without sign extension here, as the higher bits will
be used for a tag. (See encode())

* Source/JavaScriptCore/dfg/DFGAbstractHeap.h:
(JSC::DFG::AbstractHeap::Payload::Payload):

Canonical link: <a href="https://commits.webkit.org/285323@main">https://commits.webkit.org/285323@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/833b3193b8d2605f53101d1fa00d48cedb96df15

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72084 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51504 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24869 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76249 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23293 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59305 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23113 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56866 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15372 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75151 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46692 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62093 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37321 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43356 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19567 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21643 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/65220 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65268 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19927 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77929 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/71336 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16324 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19099 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65338 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16371 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62116 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64595 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15958 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12796 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6439 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/93119 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47302 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20503 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48371 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49658 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48115 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/dfg/DFGAbstractHeap.h:
```diff
@@ -123,7 +123,7 @@ class AbstractHeap {
         Payload(const void* pointer)
             : m_isTop(false)
-            , m_value(bitwise_cast<intptr_t>(pointer))
+            , m_value(reinterpret_cast<int64_t>(pointer))
```
I think `static_cast<int64_t>(reinterpret_cast<uintptr_t>(pointer));` maybe better here since
* It is more portable, working properly on both 32-bit and 64-bit systems.
* It has correct semantics, treating pointers as unsigned values, which they are by nature.
* It ensures zero extension, avoiding any risk of sign extension when casting from a smaller pointer to a 64-bit integer.

eugeneia replied:
Out of personal interest, can you explain why/how your proposed expression is more portable/correct? I was going off the following clause for reinterpret_cast on cppreference:

> 2) A pointer can be converted to any integral type large enough to hold all values of its type (e.g. to [std::uintptr_t](https://en.cppreference.com/w/cpp/types/integer)).

Which to me reads as if this should work for both 64 and 32 bit platforms since int64_t can hold all values of both 64 bit and 32 bit pointers. Maybe I am missing something?

hyjorc1 replied:
Yes, `reinterpret_cast` performs a bitwise reinterpretation. I think the proposed solution offers improves readability:

Portability: `uintptr_t` adapts to the pointer size on both 32-bit and 64-bit systems, ensuring cross-platform compatibility.
Correct Semantics: Pointers are unsigned, and `uintptr_t` is specifically designed to handle unsigned pointer values.
Safety: While `reinterpret_cast<int64_t>(pointer)` reinterprets the bit pattern without extensions, using `uintptr_t` ensures that the pointer’s value is correctly preserved, avoiding potential truncation or sign-extension issues.

Of course, if you think that's not necessary and the proposed solution introduces extra steps, then please ignore me.

justinmichaud reviewed (approved):
r=me

eugeneia committed:
[JSC] Fix DFG::AbstractHeap initialization from pointer on armv7
https://bugs.webkit.org/show_bug.cgi?id=281378

Reviewed by Justin Michaud.

The previous incantation would cause 32 bit pointers to be
sign extended into a int64_t slot. We want to widen then
value without sign extension here, as the higher bits will
be used for a tag. (See encode())

* Source/JavaScriptCore/dfg/DFGAbstractHeap.h:
(JSC::DFG::AbstractHeap::Payload::Payload):

Canonical link: https://commits.webkit.org/285323@main
